src/hotspot/share/opto/doCall.cpp

```diff
@@ -581,6 +581,23 @@ void Parse::do_call() {
   ciKlass* speculative_receiver_type = nullptr;
   if (is_virtual_or_interface) {
     Node* receiver_node             = stack(sp() - nargs);
+
+    // For osr compilation, we have no way to know which parameter is larval and which is not. We
+    // optimistically assume all parameters are non-larval. Then, a parameter is changed to larval
+    // if we encounter a store into one of its fields, or if we encounter a constructor invocation
+    // with it being the first argument
```

Comment:
The implicit larval transition has a limited life span, it begins from the new allocation and ends just before making a call to the super class constructor. 

[https://github.com/openjdk/valhalla/blob/d94c58952d2f9a0193af4cd9df66868fd0a6e3db/src/hotspot/share/opto/doCall.cpp#L830](https://github.com/openjdk/valhalla/blob/d94c58952d2f9a0193af4cd9df66868fd0a6e3db/src/hotspot/share/opto/doCall.cpp#L830
)

The only possibility to create a larval value that gets passed on to OSR buffer is by explicitly marking it as larval using Unsafe.makePrivateBuffer API before the loop.

May I request you to kindly add a new or identify an existing test case for this scenario

Reply:
> The only possibility to create a larval value that gets passed on to OSR buffer is by explicitly marking it as larval using Unsafe.makePrivateBuffer API before the loop.

Why is that? We can have a loop triggering OSR within that limited life span, right?

Reply:
Hi @TobiHartmann ,
The case mentioned above has the value type constructor call out of loop,  lifespan of larval is limited till construction.   
For OSR compilation, we begin compilation starting from loop entry and generate the downstream graph.

In practice, an OSR compiled method should always receive a value in non-larval state in its argument buffer.

The case I am referring to is the one where we have an Unsafe.makePrivateBuffer() call outside the loop to explicitly transition value instance to larval state,  if this mutable value is used within the loop then OSR buffer should hold the value in larval state.

Reply:
> In practice, an OSR compiled method should always receive a value in non-larval state in its argument buffer.

@jatin-bhateja This is wrong, if you look at the constructor of `TestValueConstruction.MyValue10`

    public MyValue10(int x, int cnt) {
        this.x = x;
        int res = 0;
        for (int i = 0; i < cnt; ++i) {
            res += x;
        }
        this.y = res;
        super();
    }

Then `this` is a larval value type until the end of `super()`. As a result, an OSR compilation starting at the loop will see a larval value object in its argument buffer.

Reply:
Please note that while `javac` often emits the constructor call right after the `new` instruction. There is no restriction and it can insert arbitrary things between them. As a result, the Hotspot compiler must be able to deal with arbitrary bytecode shapes, not only those emitted by `javac`